### ISSUE TITLE
maximum poll interval (90000ms) exceeded by 846018558ms

### ISSUE THREAD

**Issue by WhiteWind** [edenhill/librdkafka#3537]
===========
High-level consumer sometimes fails with error:
Application maximum poll interval (300000ms) exceeded by 846018558ms: Local: Maximum application poll interval (max.poll.interval.ms) exceeded

846018s = 235h = 9days, but application was started a few hours ago.

I don't know what debug=... setting I should set. App may run without issue for a few days, so debug=all will produce too much output

Checklist
=========

Please provide the following information:

 - [x] librdkafka version v1.7.0
 - [x] Apache Kafka version: 2.3.1
 - [x] librdkafka client configuration: api.version.request= true,auto.offset.reset=earliest,enable.auto.commit=false,fetch.message.max.bytes= 50000,heartbeat.interval.ms=10000,message.max.bytes= 17000000,queued.max.messages.kbytes= 17000,session.timeout.ms=30000,reconnect.backoff.max.ms=5000,reconnect.backoff.ms=200,retry.backoff.ms=200
 - [x] Operating system: Windows Server 2016 standard
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
I'm guessing your system clock was changed.

**Comment by WhiteWind** [comment]
Maybe. But it's just native Windows time sync service on virtual machine.
It does change time by jumps, not by adjustments, so it may introduce negative time offsets.

Isn't it better to calculate timeouts using GetTickCount64/CLOCK_MONOTONIC_COARSE instead of wall clock time ?

**Comment by WhiteWind** [comment]
I see, that you use CLOCK_MONOTONIC on Linux and QPC on Windows. My problem is that my VM does not have an invariant TSC. 
Is such precision really necessary? Wouldn't GetTickCount64 be enough?

**Comment by edenhill** [comment]
We need sub-millisecond granularity for some timers, and it'd be a headache to split up into two different implementations.
GetTickCount64 seems to have a granularity of 10-16ms, which is not good enough:
https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-gettickcount64

**Comment by WhiteWind** [comment]
Ok, I see. 
May there be a bug with integer overflow:

```exceeded = (int)((rd_clock() - last_poll) / 1000ll) -
                rk->rk_conf.max_poll_interval_ms;

        if (unlikely(exceeded > 0))
                return exceeded;```

or it's just TSC difference is so huge (9 days) on different cores?

**Comment by edenhill** [comment]
Could you add some debug printfs to ..check_tmr_cb()? E.g., print both rd_clock() and last_poll?

**Comment by WhiteWind** [comment]
Yes I can, but this issue is present only on production server, so testing will take some time

**Comment by WhiteWind** [comment]
Got it:

> [MAXPOLL][thrd:main]: Last poll: 844135975 Now: 2147483647
> [MAXPOLL][thrd:main]: Application maximum poll interval (180000ms) exceeded by 2065699270ms

**Comment by WhiteWind** [comment]
Oh, probably I used wrong format specifier...

**Comment by edenhill** [comment]
Right, use `"Last poll: %"PRId64", Now: %"PRId64`

**Comment by WhiteWind** [comment]
[MAXPOLL][thrd:main]: Last poll: 9223372034273984178 Now: 555764956851
[MAXPOLL][thrd:main]: Application maximum poll interval (180000ms) exceeded by 2069994237ms

**Comment by edenhill** [comment]
That last_poll doesn't seem right
`	9223372034273984178 /* 0x7fffffff662c3eb2 */`

Can you show your code?

**Comment by WhiteWind** [comment]
This is what I have changed from git tag v1.7.0

```
diff --git a/src/rdkafka_int.h b/src/rdkafka_int.h
index a37e61cc..56a2367f 100644
--- a/src/rdkafka_int.h
+++ b/src/rdkafka_int.h
@@ -981,6 +981,7 @@ rd_kafka_resp_err_t rd_kafka_subscribe_rkt (rd_kafka_topic_t *rkt);
 static RD_INLINE RD_UNUSED int
 rd_kafka_max_poll_exceeded (rd_kafka_t *rk) {
         rd_ts_t last_poll;
+               rd_ts_t now_c;
         int exceeded;

         if (rk->rk_type != RD_KAFKA_CONSUMER)
@@ -993,11 +994,17 @@ rd_kafka_max_poll_exceeded (rd_kafka_t *rk) {
         if (last_poll == INT64_MAX)
                 return 0;

-        exceeded = (int)((rd_clock() - last_poll) / 1000ll) -
+               now_c = rd_clock();
+        exceeded = (int)((now_c - last_poll) / 1000ll) -
                 rk->rk_conf.max_poll_interval_ms;

-        if (unlikely(exceeded > 0))
-                return exceeded;
+               if (unlikely(exceeded > 0)) {
+                       rd_kafka_log(rk, LOG_WARNING, "MAXPOLL",
+                               "Last poll: %I64d "
+                               "Now: %I64d",
+                               last_poll, now_c);
+                       return exceeded;
+               }

         return 0;
 }
diff --git a/src/win32_config.h b/src/win32_config.h
index 4579cf70..a1a00c5a 100644
--- a/src/win32_config.h
+++ b/src/win32_config.h
@@ -39,7 +39,7 @@
 #define WITH_ZSTD 1
 /* zstd is linked dynamically on Windows, but the dynamic library provides
  * the experimental/advanced API, just as the static builds on *nix */
-#define WITH_ZSTD_STATIC 1
+#define WITH_ZSTD_STATIC 0
 #define WITH_SASL_SCRAM 1
 #define WITH_SASL_OAUTHBEARER 1
 #define ENABLE_DEVEL 0
@@ -49,6 +49,6 @@
 #define SOLIB_EXT ".dll"

 /* Notice: Keep up to date */
-#define BUILT_WITH "SSL ZLIB SNAPPY SASL_SCRAM PLUGINS HDRHISTOGRAM"
+#define BUILT_WITH "SSL ZLIB SNAPPY SASL_CRAM PLUGINS HDRHISTOGRAM"

 #endif /* _RD_WIN32_CONFIG_H_ */
```

**Comment by edenhill** [comment]
That value for last_poll is very weird.
Could you add something like:
```c
if (rd_clock() < 0 || rd_clock() > (1 << 62))
  RD_BUG("rd_clock() is %"PRId64);
```
to rd_kafka_app_polled()?

**Comment by WhiteWind** [comment]
Library is 32bit. `rd_atomic64_set()` is using `InterlockedExchange64()`, but `rd_atomic64_get()` is not using any Interlocked function. MSDN says:
> The interlocked functions provide a simple mechanism for synchronizing access to a variable that is shared by multiple threads. This function is atomic with respect to calls to **other interlocked functions**.

Probably this is it

**Comment by edenhill** [comment]
Oh, good find!

**Comment by WhiteWind** [comment]
Now I'm testing with 
```
static RD_INLINE int64_t RD_UNUSED rd_atomic64_get(rd_atomic64_t *ra) {
#if defined(_WIN32) || defined(__SUNPRO_C)
	return InterlockedCompareExchange64(&ra->val, 0, 0);
#elif !HAVE_ATOMICS_64
```

**Comment by edenhill** [comment]
Let me know how it goes, and if successful, would you care to submit a PR?

**Comment by WhiteWind** [comment]
Almost 24 hours without such error, submitted a PR

**Comment by edenhill** [comment]
Fixed by @WhiteWind in #3815
